lxd/storage/drivers/driver_btrfs_volumes.go

```diff
@@ -1219,7 +1219,12 @@ func (d *btrfs) MigrateVolume(vol Volume, conn io.ReadWriteCloser, volSrcArgs *m
 	}
 
 	// Generate restoration header, containing info on the subvolumes and how they should be restored.
-	migrationHeader, err := d.restorationHeader(vol, volSrcArgs.Snapshots)
+	snapshots, err := d.volumeSnapshotsSorted(vol, op)
```

Comment:
@monstermunchkin please can you elaborate (and importantly add a comment for future readers) the commit message:

"The passed snapshots dont't always include all which need to be
transferred."

Why don't they always?

Reply:
Consider the following example:

```
lxc storage volume create l1:btrfs vol1
lxc storage volume snapshot l1:btrfs vol1
lxc storage volume snapshot l1:btrfs vol1
lxc storage volume snapshot l1:btrfs vol1 snapremove

lxc storage volume copy l1:btrfs/vol1 l2:btrfs/vol2 --refresh

lxc storage volume create l1:btrfs vol3
lxc storage volume snapshot l1:btrfs vol3
lxc storage volume snapshot l1:btrfs vol3
lxc storage volume snapshot l1:btrfs vol3
lxc storage volume copy l1:btrfs/vol3 l2:btrfs/vol2 --refresh
```

The last command will fail with

    Error: Failed storage volume creation: Error transferring storage volume: Failed to run: btrfs receive -e /home/thomas/work/lxd/server/storage-pools/btrfs/custom/migration.1597063747/snap2: ERROR: cannot find parent subvolume

On the source we have
```
$ lxc storage volume list l1:btrfs 
+-------------------+------------------------------------------------------------------+-------------+--------------+---------+
|       TYPE        |                               NAME                               | DESCRIPTION | CONTENT-TYPE | USED BY |
+-------------------+------------------------------------------------------------------+-------------+--------------+---------+
| custom            | vol1                                                             |             | filesystem   | 0       |
+-------------------+------------------------------------------------------------------+-------------+--------------+---------+
| custom (snapshot) | vol1/snap0                                                       |             | filesystem   | 0       |
+-------------------+------------------------------------------------------------------+-------------+--------------+---------+
| custom (snapshot) | vol1/snap1                                                       |             | filesystem   | 0       |
+-------------------+------------------------------------------------------------------+-------------+--------------+---------+
| custom (snapshot) | vol1/snapremove                                                  |             | filesystem   | 0       |
+-------------------+------------------------------------------------------------------+-------------+--------------+---------+
| custom            | vol3                                                             |             | filesystem   | 0       |
+-------------------+------------------------------------------------------------------+-------------+--------------+---------+
| custom (snapshot) | vol3/snap0                                                       |             | filesystem   | 0       |
+-------------------+------------------------------------------------------------------+-------------+--------------+---------+
| custom (snapshot) | vol3/snap1                                                       |             | filesystem   | 0       |
+-------------------+------------------------------------------------------------------+-------------+--------------+---------+
| custom (snapshot) | vol3/snap2                                                       |             | filesystem   | 0       |
+-------------------+------------------------------------------------------------------+-------------+--------------+---------+
```

... and on the target we have
```
$ lxc storage volume list l2:btrfs 
+-------------------+------------+-------------+--------------+---------+
|       TYPE        |    NAME    | DESCRIPTION | CONTENT-TYPE | USED BY |
+-------------------+------------+-------------+--------------+---------+
| custom            | vol2       |             | filesystem   | 0       |
+-------------------+------------+-------------+--------------+---------+
| custom (snapshot) | vol2/snap0 |             | filesystem   | 0       |
+-------------------+------------+-------------+--------------+---------+
| custom (snapshot) | vol2/snap1 |             | filesystem   | 0       |
+-------------------+------------+-------------+--------------+---------+
```

The LXD backend will see that the target is missing the `snap2` snapshot, and will therefore only include `snap2` in `volSrcArgs.Snapshots`. In reality though, `vol2/snap0` and `vol2/snap1` are copies from `vol1/snap0` and `vol1/snap1` respectively, not those of `vol3`. Therefore, they need to be copied as well. The btrfs driver can and will detect this using the migration headers  and do the correct thing which is to copy the entire volume (including all snapshots).

@tomponline does that make sense?

Reply:
> Hrm yeah I see what you're saying. This sounds like its a driver-specific solution to this problem https://github.com/lxc/lxd/issues/10436 right?

I believe so.

> Eventually we will need to fix it for all storage drivers by using a volatile uuid in the config to detect syncing different volumes independent of the actual storage layer.

That would make sense.

> Do we not also need similar for ZFS?

zfs handles this already.

Reply:
> zfs handles this already.

Ah right, in the same way? Please could you link to it for me? Thanks